pkg/storage/etcd3/v3_helper.go

```diff
@@ -0,0 +1,333 @@
+	}
+
+	return nil
+}
+
+func (h *v3Helper) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+	getResp, err := h.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	// if not found:
+	//   if ignoreNotFound = true: set objPtr to be zero value
+	//   else: return not found error
+	if len(getResp.Kvs) == 0 {
+		if ignoreNotFound {
+			return setZeroValue(out)
+		} else {
+			return storage.NewKeyNotFoundError(key, 0)
+		}
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	if version, err := h.versioner.ObjectResourceVersion(obj); err == nil && version != 0 {
+		return errors.New("resourceVersion should not be set on objects to be created")
+	}
+	data, err := runtime.Encode(h.codec, obj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	// If rev = 0, it means the key doesn't exist.
+	resp, err := h.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", 0),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !resp.Succeeded {
+		return storage.NewKeyExistsError(key, 0)
+	}
+
+	if out != nil {
+		putResp := resp.Responses[0].GetResponsePut()
+		return decode(h.codec, h.versioner, data, out, putResp.Header.Revision)
+
+	}
+	return nil
+}
+
+func (h *v3Helper) Delete(ctx context.Context, key string, out runtime.Object) error {
+	if _, err := conversion.EnforcePtr(out); err != nil {
+		panic("unable to convert output object to pointer")
+	}
+
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+	// We need to do get and delete in single transaction in order to
+	// know the value and revision before deleting it.
+	txnResp, err := h.client.KV.Txn(ctx).If().Then(
+		clientv3.OpGet(key),
+		clientv3.OpDelete(key),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	getResp := txnResp.Responses[0].GetResponseRange()
+	if len(getResp.Kvs) == 0 {
+		// To be compatible with original logic of etcdHelper.
+		// In etcd2, delete non-existing key will return error of "key not found"
+		return storage.NewKeyNotFoundError(key, 0)
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) GetToList(ctx context.Context, key string, filter storage.FilterFunc, listObj runtime.Object) error {
+	listPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	getResp, err := h.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+	if len(getResp.Kvs) == 0 {
+		return nil
+	}
+	// Make a list off of the single returned value.
+	values := append([][]byte(nil), getResp.Kvs[0].Value)
+	revs := append([]uint64(nil), uint64(getResp.Kvs[0].ModRevision))
```

Comment:
I'm slightly against having two arrays with the implicit assumption that their length are equal.
Can you please introduce a type, e.g.

type elemForDecode struct {
  data []byte
  rev   uint64
}
[please come up with better name for it]
and use it instead of these two arrays?